src/NHibernate/Linq/Visitors/ResultOperatorProcessors/ProcessAggregate.cs

```diff
@@ -19,20 +19,22 @@ public void Process(AggregateResultOperator resultOperator, QueryModelVisitor qu
 				resultOperator.Func.Parameters[0],
 				paramExpr);
 
-			var inputList = Expression.Parameter(typeof(IEnumerable<>).MakeGenericType(typeof(object)), "inputList");
-
-			var castToItem = ReflectionCache.EnumerableMethods.CastDefinition.MakeGenericMethod(new[] { inputType });
-			var castToItemExpr = Expression.Call(castToItem, inputList);
-
+			// NH-3850: changed from list transformer (working on IEnumerable<object>) to post execute
+			// transformer (working on IEnumerable<inputType>) for globally aggregating polymorphic results
+			// instead of aggregating results for each class separately and yielding only the first.
+			// If the aggregation relies on ordering, final result will still be wrong due to
```

Comment:
Test added. By the way, hardened the non mutable seed test for it to be no more sensible to ordering. I prefer avoid relying on current polymorphic classes ordering whenever possible for tests, since it does look undocumented.